best_practices.md

```diff
@@ -0,0 +1,127 @@
+
+### Motivation
+- **User Experience**: Provides more informative error messages, improving user experience.
+- **Debugging**: Easier to debug issues with clear assertions and fatal errors.
+- **Performance**: Avoids the overhead associated with exception handling.
+
+## 6. Avoid `std::move` When Returning from Functions
+
+### Practice
+Don't use `std::move` when returning from a function, as it breaks Return Value Optimization (RVO).
+
+### Explanation
+RVO is an optimization technique that eliminates unnecessary copying or moving of objects when they are returned from a function. Using `std::move` can prevent the compiler from applying RVO, resulting in less efficient code.
+
+### Motivation
+- **Performance**: Ensures that RVO can be applied, minimizing unnecessary copying or moving of objects.
+- **Efficiency**: Maintains optimal performance by allowing the compiler to perform return value optimizations.
+
+## 7. Avoid `const T&&` or `const auto&&` and Returning `const` Values
+
+### Practice
+Never use `const T&&` or `const auto&&`, and never return `const` values from functions.
+
+### Explanation
+Using `const` rvalue references (`const T&&` or `const auto&&`) and returning `const` values from functions can prevent the compiler from optimizing the code. These practices block the use of move constructors, which are crucial for efficient resource management and performance optimization.
+
+### Motivation
+- **Performance**: Enables the use of move constructors, enhancing performance.
+- **Optimization**: Allows the compiler to optimize code more effectively.
+- **Efficiency**: Facilitates efficient resource management by leveraging move semantics.
+
+  ## 8. Use `std::move` with `emplace_back` and Prefer `push_back` When Appropriate
+
+### Practice
+The `emplace_back` call takes the same parameters as the constructor. Passing a type by value calls the copy constructor, negating the benefits of `emplace_back`. Use `std::move` when necessary, and consider using `push_back` with the same efficiency.
+
+### Explanation
+`emplace_back` constructs an element in place, avoiding unnecessary copies or moves when the arguments match the constructor parameters. However, if parameters are passed by value, a copy constructor is invoked. Using `std::move` ensures that resources are moved rather than copied. In some cases, `push_back` can be just as efficient and more straightforward.
+
+### Motivation
+- **Efficiency**: Ensures resources are moved, not copied, maintaining performance.
+- **Clarity**: Using `push_back` can be simpler and equally efficient in certain scenarios.
+- **Optimization**: Maximizes the benefits of in-place construction and move semantics.
+
+## 9. Move constructor should not throw. Mark Move Constructors as `noexcept`
+
+### Practice
+Move constructors should be marked as `noexcept`, otherwise many STL optimizations will not be applied.
```

Comment:
This one is really important and too easy to overlook. Do you know if clang-tidy has any rule to catch this?

Reply:
clang tidy is the next step 100%.